custom/aaa/models.py

```diff
@@ -1212,3 +1207,70 @@ def _dictfetchone(cursor):
         return ret[0]
 
     return {}
+
+
+class DenormalizedAWC(models.Model):
+    domain = models.TextField()
+    state_id = models.TextField()
+    district_id = models.TextField()
+    block_id = models.TextField()
+    supervisor_id = models.TextField()
+    awc_id = models.TextField(unique=True)
+
+    @classmethod
+    def build(cls, domain):
+        cls.objects.filter(domain=domain).delete()
+        domain_locations = SQLLocation.objects.filter(domain=domain).values(
+            'pk', 'parent_id', 'location_id', 'location_type__code')
+        locations_by_pk = {
+            loc['pk']: loc
+            for loc in domain_locations
+        }
+        awcs = [loc for loc in domain_locations if loc['location_type__code'] == 'awc']
+        to_save = []
+        for awc in awcs:
+            loc = DenormalizedAWC(domain=domain, awc_id=awc['location_id'])
+
+            current_location = awc
+            while current_location['parent_id']:
+                current_location = locations_by_pk[current_location['parent_id']]
+                attr_name = '{}_id'.format(current_location['location_type__code'])
+                setattr(loc, attr_name, current_location['location_id'])
+
+            to_save.append(loc)
+
+        cls.objects.bulk_create(to_save)
+
+
+class DenormalizedVillage(models.Model):
+    domain = models.TextField()
+    state_id = models.TextField()
+    district_id = models.TextField()
+    taluka_id = models.TextField()
+    phc_id = models.TextField()
+    sc_id = models.TextField()
+    village_id = models.TextField(unique=True)
+
+    @classmethod
+    def build(cls, domain):
```

Comment:
These two functions were just different enough that I didn't make a helper to make it more DRY.